conn.go

```diff
@@ -120,10 +146,18 @@ func (c *Conn) printLine(cmd string, args ...interface{}) {
 }
 
 func (c *Conn) readResp(r req, readBody bool, f string, a ...interface{}) (body []byte, err error) {
+	readTimeout := c.readTimeout
+	// For reserve-with-timeout commands, add reserve time to read timeout.
+	if r.op == "reserve-with-timeout" {
+		readTimeout = c.readTimeout + c.reserveTimeout
+	}
+
 	c.c.StartResponse(r.id)
 	defer c.c.EndResponse(r.id)
+	c.netConn.SetReadDeadline(time.Now().Add(readTimeout))
 	line, err := c.c.ReadLine()
 	for strings.HasPrefix(line, "WATCHING ") || strings.HasPrefix(line, "USING ") {
+		c.netConn.SetReadDeadline(time.Now().Add(readTimeout))
```

Comment:
I strongly suspect setting deadlines like this is going to lead to the connection getting out of sync, leading to things like:
 - Portions of the response for one command being interpreted as the first line of the response for a subsequent one.
 - The response for one command being interpreted as the response for the next one because the original one timed out.

Fundamentally, I think the wire protocol for beanstalkd uses too much implicit state to safely support such deadlines.

As an example:
Let's say there's a network blip so, the response to a put gets delayed (or equivalently from the client's side, the request is delayed on its way to beanstalkd).

In that case, as this is written, the initial request will timeout, and the application gets back:
`ConnError{c, "put", sometimeoutErr}` (and job ID 0)

In the mean time, the response makes it to our client, with, say ID `523`.

However, most applications don't put just one job, so the next `put` now comes along and the client now gets to this line again; this time, there's a response waiting for it.
If the tube is the same, the client skips the `use` line and goes directly to a `put`.

 - This time, it doesn't matter how fast that request is, the client will use the response that's queued from the previous request, rather than the one directly associated with the response.

Reply:
Hi @dfinkel thanks for your feedback. 

As I understand it, the Go conn package requires that the read timeout be set before each read, this is why it is in the form of an absolute deadline in time, rather than a duration.

If reading the response times out, then as I understand it the connection itself is torn down (this is certainly my intention and experience of the behaviour), so that no other commands can be sent or read from the connection, and a new one must be established.

So I can appreciate what you're saying here, but I don't think it will actually happen that way, as the read will return with a ConnError

https://github.com/beanstalkd/go-beanstalk/pull/26/files#diff-a2ddca1600f34abb91244424f8431c2cL130

The alternative is that a network glitch causes the TCP connection to hang and your process will wait indefinitely for a response (we've experienced this in real life, which is what prompted me to add this PR).

Reply:
You are correct that the go `net.Conn` interface requires that the timeout be set to an absolute value. However, the connection is definitely not torn down if it fails.

The only effects are: all subsequent reads (or writes if that deadline is set) will fail with an i/o timeout error until the deadline is updated again.

The documentation on [`net.Conn.SetTimeout()` documentation](https://golang.org/pkg/net/#Conn) alludes to this when it mentions that this can be used for idle polling.

The read will return a ConnError, but that's not always something that should be considered fatal.
 
e.g. The `beanstalk.ConnError` type wraps `ErrTimeout` and `ErrDeadline` which are non-fatal and just reflect either: 1) there is another job whose ttr is about to expire, or 2) the timeout on a `reserve-timeout` call (any `Reserve()` call in this client) has expired. Neither should trigger a connection tear-down.

I appreciate that some lost packets can cause the client to hang indefinitely as-is. In a request handling context where one is only using `Put` to push jobs, it even makes sense to have deadlines and kill the connection on most errors. However, this is a generic library that handles both producers and consumers. The beanstalk protocol is fundamentally not designed for this, and as-is the client will end up in an undefined state after a timeout, such that it will return garbage results to the client.

If you intend for it to close the connection on timeout, you'll have to add a `c.Close` and document it.

With that said, I think a better first step would be to constrain the implementation to using `*net.TCPConn` (or at least an interface that supports its [`SetKeepAlive()`](https://golang.org/pkg/net/#TCPConn.SetKeepAlive) and [`SetKeepAlivePeriod()`](https://golang.org/pkg/net/#TCPConn.SetKeepAlivePeriod)).
This way if the connection gets closed on the other side during a network blip, the connection can actually get a reset packet once the host is reachable, which will kill the connection, causing it to return `io.EOF` on any hanging request.

Granted, this is incomplete protection against network blackholes, but it does avoid adding a new poisoned state to the connection.